draft-specifications/Locators.md

```diff
@@ -0,0 +1,137 @@
+on, it encodes the transport  protocol name, key and other additional data which
+would be used by any given  netlayer to reach the node.
+
+# Additional Documents
+
+This document does not stand alone, it relies upon multiple other documents
+which together build up OCapN  (Object Capability Network) specifications.
+
+This specification uses the following other specifications:
+
+-   [Syrup](): The serialization format used to encode the in band
+    representation of the locator. 
+-   [OCapN Netlayers](): Specification to open a secure communication channel
+    between two sessions, often on  different networks.
+-   [RFC3986: Uniform Resource Identifier (URI): Generic
+    Syntax](https://www.rfc-editor.org/rfc/rfc3986): For encoding URI
+    representations of locations.
+
+# [Node Locator](#ocapn-node)
+
+This identifies an OCapN node, not a specific object. This includes enough
+information to specify which netlayer and provide that netlayer with all of the
+information needed to create a  bidirectional channel to that node.
+
+The node locator include the following pieces of information (more details
+below):
+
+- **Designator**: Usually representing the key, however can be any value
+  determined by the netlayer
+- **Transport**: A unique identifier to specify a netlayer
+- **Hints**: A hashmap of additional connection information.
+
+When comparing two node locators, the designator and transport are the only
+pieces of information which need to match. Two node locators can have the same
+designator and transport but  different hints and be considered to be the same
+node.
+
+## [Syrup Serialization](#node-syrup-serialization)
+
+It's encoded as a record with the label `ocapn-node` and three arguments:
+
+```
+<ocapn-node designator  ; string
+            transport   ; symbol (cannot contain ".")
+            hints>      ; hashmap | false
+```
+
+### Hints
+
+This is a hashmap of key and values which are designed to encode additional
+connection information that the  netlayer might need to reach the node. The keys
+should be symbols with the values as strings.
+
+There can be any number of hints, including none at all. If no hints are used
+this field should be set to false.
+
+## URI Serialization
+
+The URI serialization is a URI form designed to be provided to users to be given
+out of band to bootstrap a CapTP session.
+
+The URI format is as follows:
+```
+;; Without hints (i.e. hints are set to false)
+ocapn://<designator>.<transport>
+
+;; With hints
+ocapn://<designator>.<transport>?hint1=value1&hint2=value2
+```
+
+This is a URI with the scheme `ocapn` followed by the designator, a `.` and then
+the transport name. If any  hints exist they're added as part of the query
+parameters, otherwise emitted. 
+
+Note that the designator permits `.` to be used within it, however the final `.`
+should designate the seperator  between the designator segment and the transport
+identifier.
+
```

Comment:
Can the transport bit be moved into the schema part of an URI?

That way the semantics will precede designator, one can sort and prefix filter URIs by transport without having to parse the designator, and the generic rfc3986 facilities can be used.

- `ocapn+tor://hsdfhsdjkhflasdkhf.onion:12345?hints...`
- `ocapn+tcp://[fd7a:115c::9f1b]:12345?hints...`
- `ocapn+cadet://secret@V8XXK9GAN5ZJFRFQP8MQX3D83BZTSBQVHKWWD0JPE63Z821906EG`

Reply:
Yeah, this has actually been discussed a bit in https://github.com/ocapn/ocapn/issues/29 and when designing these URIs in Goblins one of the contenders we (spritely) were considering (even though we decided not to go with it).

Reply:
I would move that the transport move the other direction, such that client can filter connection hints for supported transports. Concretely:

```
[ocapn]
key: b0b5c0ffeefacade
nonce: deadb33f
[hints]
- tor://hsdfhsdjkhflasdkhf.onion
- tcp://[fd7a:115c::9f1b]:12345
- cadet://secret@V8XXK9GAN5ZJFRFQP8MQX3D83BZTSBQVHKWWD0JPE63Z821906EG
```